### ISSUE TITLE
Expose `objects` factory to `Settings` scripts

### ISSUE THREAD

**Issue by bamboo** [gradle/gradle#14171]
With `Settings` plugins and `Settings` extensions becoming more popular it is expected that more build and configuration logic will migrate to settings scripts.

And while it is possible to access an `ObjectFactory` service from a settings script, it is cumbersome, relies on non public API for service lookup and the provided `ObjectFactory` is not able to inject the most commonly requested services.

The closest replacement is `extensions.create` but it has the side-effect of registering the extension with the `Settings` object which is not always desirable.

For perspective, here's how one could add a configuration cache friendly background action to the Gradle Enterprise plugin configuration Today using only public API:

```kotlin
import com.gradle.scan.plugin.BuildScanExtension

plugins {
    id("com.gradle.enterprise") version("3.4")
}

open class MyBackgroundAction @javax.inject.Inject constructor(
    val rootDir: File,
    val execOperations: ExecOperations
) : Action<BuildScanExtension> {

    override fun execute(buildScan: BuildScanExtension) {
        println("rootDir: $rootDir")
        println("execOperations: $execOperations")
    }
}

gradleEnterprise {
    buildScan.background(
        extensions.create<MyBackgroundAction>("myBackgroundAction", rootDir)
    )
}
```

**Comment by grossws** [comment]
Is `serviceOf<ObjectFactory>()` a sensible workaround?

**Comment by bamboo** [comment]
Hey @ducrohet, we'll look into adding the missing API soon, in the meantime, regular settings plugins can use service injection:

```kotlin
                open class SettingsPlugin @javax.inject.Inject constructor(
                    val objectFactory: ObjectFactory
                ) : Plugin<Settings> {
                    override fun apply(settings: Settings) {
                    }
                }
```

**Comment by ducrohet** [comment]
Yes, that's what we just ended up doing.

**Comment by yogurtearl** [comment]
the workaround in https://github.com/gradle/gradle/issues/14171#issuecomment-978783602 seems to work fine as well.
Looks like these work fine directly in `settings.gradle.kts`

```kotlin
serviceOf<ExecOperations>()
val objects = serviceOf<ObjectFactory>()
```

**Comment by mlopatkin** [comment]
This is going to be solved as part of a more general #13121.
